src/batch/src/executor2/top_n.rs

```diff
@@ -127,48 +128,52 @@ impl TopNExecutor {
             },
             child,
             identity,
+            chunk_size
         }
     }
 }
 
-#[async_trait::async_trait]
-impl Executor for TopNExecutor {
-    async fn open(&mut self) -> Result<()> {
-        self.child.open().await?;
-
-        while let Some(chunk) = self.child.next().await? {
-            self.top_n_heap.fit(Arc::new(chunk));
-        }
-        self.child.close().await?;
-
-        Ok(())
+impl Executor2 for TopNExecutor2 {
+    fn schema(&self) -> &Schema {
+        self.child.schema()
     }
 
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
-        if let Some(chunk) = self.top_n_heap.dump() {
-            Ok(Some(chunk))
-        } else {
-            Ok(None)
-        }
+    fn identity(&self) -> &str {
+        &self.identity
     }
 
-    async fn close(&mut self) -> Result<()> {
-        Ok(())
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
     }
+}
 
-    fn schema(&self) -> &Schema {
-        self.child.schema()
-    }
+impl TopNExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        #[for_await]
+        for data_chunk in self.child.execute() {
+            let data_chunk = data_chunk?;
+            self.top_n_heap.fit(Arc::new(data_chunk));
+        }
 
-    fn identity(&self) -> &str {
-        &self.identity
+        if let Some(data_chunk) = self.top_n_heap.dump() {
+            let rows = data_chunk.rows().collect::<Vec<_>>();
+            let mut chunks = rows.chunks(self.chunk_size); 
+            while let Some(chunk) = chunks.next(){
+                let rows = chunk.iter().map(|rowref |rowref.to_owned_row()).collect::<Vec<_>>();
+                let ret_chunk = 
```

Comment:
I have some question here , the rust compiler throw this error

> risingwave_batch::executor2::Executor2
> pub fn schema(&self) -> &[Schema]()
> Returns the schema of the executor's return data.
> 
> Schema must be available before init.
> 
borrow of moved value: `self.child`
> move occurs because `self.child` has type `Box<dyn Executor2>`, which does not implement the `Copy` traitrustc[E0382](https://doc.rust-lang.org/error-index.html#E0382)
> [top_n.rs(154, 38): ]()`self.child` moved due to this method call
> [mod.rs(59, 16): ]()this function takes ownership of the receiver `self`, which moves `self.child`

why[ there](https://github.com/singularity-data/risingwave/blob/2027b1f1686eee2da853af2ba1edc2edcf3f1d3b/src/batch/src/executor2/filter.rs#L52) is no error?

Reply:
I think this is because `self.child.execute()` already took ownership of `self.child`, and you use `self.child.schema()` after that.

Reply:
oop...I forget that execute() take the ownership of the child. I will fix this